### PR TITLE
Arc 778 - Change generated create pull request link to new format to allow title overrides with query params

### PR DESCRIPTION
**Pull request by joshkay10**
Change PR URL to new format to include title override query param

e.g. from:
https://github.com/atlassian/github-for-jira/pull/new/branch-x
to
https://github.com/atlassian/github-for-jira/compare/branch-x?title=ARC-778-branch-x

This URL was created in 3 seperate locations for the attribute **createPullRequestUrl**
- github-for-jira/src/transforms/pull-request.ts
- github-for-jira/src/transforms/branch.ts
- github-for-jira/src/sync/transforms/branch.ts

So I created a single Util function to import into each of the above and call function **generateCreatePullRequestUrl** which will generate the URL and ensure that the issue keys are attached appropriately.

The above function has unit tests for with single, multiple and none issue keys here
-  github-for-jira/test/unit/transforms/util/pullRequestLinkGenerator.test.ts

### CODE REVIEW

File: src/transforms/util/pullRequestLinkGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * Generates a create pull request link for GH in the format [URL]/compare/[name]?title=[...keys]-[name]
+ * This format allows the use of title query param to set the PR name
+ */
+export const generateCreatePullRequestUrl = (baseUrl: string, name: string, issueKeys: string[] | null) => {
```
Comment (KAbakumovAtl) [inline src/transforms/util/pullRequestLinkGenerator.ts:5]: /nit should we do `issueKeys?: string[]` instead?

File: test/fixtures/api/graphql/branch-ref-nodes.json [hunk 2]
```diff
@@ -13,7 +13,7 @@
                   }
                 ]
               },
-              "name": "TES-321-branch-name",
+              "name": "test-branch-name",
```
Comment (KAbakumovAtl) [inline test/fixtures/api/graphql/branch-ref-nodes.json:16]: I wonder in which tests it is used? I am afraid it can affect some tests.

(/nit Should we name it explicitly `branch-with-issue-key-in-the-last-commit` to improve tests readability?)

File: test/unit/transforms/util/pullRequestLinkGenerator.test.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { generateCreatePullRequestUrl } from "../../../../src/transforms/util/pullRequestLinkGenerator";
+
+describe("pullRequestLinkGenerator()", () => {
+
+	it("should structure the url correctly with issue keys in title - for single issue key", async () => {
+		const BASE_URL = "github";
+		const NAME = "branch-7";
+		const ISSUE_KEYS = ["TEST-222"];
+		const ISSUE_KEYS_FLAT = "TEST-222"
+		const EXPECTED_RESULT = `github/compare/${NAME}?title=${ISSUE_KEYS_FLAT}%20-%20${NAME}`;
```
Comment (KAbakumovAtl) [inline test/unit/transforms/util/pullRequestLinkGenerator.test.ts:10]: Let's use the exact string in the Expected Result. 

Otherwise this test repeats the implementation

File: src/transforms/util/pullRequestLinkGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+	// Remove duplicate keys
+	const uniqueIssueKeys: string[] = [...new Set(issueKeys)];
+	let title: string | undefined = uniqueIssueKeys?.join(" ");
+
+	// Only prepend issue keys if they exist
+	if (title) {
+		title = encodeURIComponent(`${title} - ${name}`);
+	} else {
+		title = encodeURIComponent(`${name}`);
+	}
```
Comment (mboudreau) [inline src/transforms/util/pullRequestLinkGenerator.ts:15]: I think this whole thing could be simplified to 
```
const keys = issueKeys.length ? `${issueKeys.join(" ")} - ` : "";
const title = encodeURIComponent(keys + name);
```

File: src/transforms/util/pullRequestLinkGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+	return  `${baseUrl}/compare/${name}?title=${title}`;
```
Comment (mboudreau) [inline src/transforms/util/pullRequestLinkGenerator.ts:16]: You'll need to `encodeURIComponent` of name here.

An idea I had for this since you're doing url params is you could use the `URLParamsMiddleware` function to do what you want:

```
return urlParamsMiddleware({
   url: "{baseUrl}/compare/{name}?title={title}"
   urlParams: {
      baseUrl,
      name,
      title
   }
}).url;
```

Might be overkill though.  I think we might need to create a shared util from the logic in URLParamsMiddleware that does this for any url.

**Review by mboudreau — CHANGES_REQUESTED** [review]

File: src/transforms/util/pullRequestLinkGenerator.ts (hunk 5 above)
Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts:9]: After the changes from your previous comments the function is quite succinct, so will omit the urlParamsMiddleware at this stage.

File: src/transforms/pull-request.ts [hunk 6]
```diff
@@ -82,7 +82,7 @@ export default async (github: GitHubAPI, pullRequest: Octokit.PullsGetResponse,
 				? []
 				: [
 					{
-						createPullRequestUrl: generateCreatePullRequestUrl(pullRequest?.head?.repo?.html_url, pullRequest?.head?.ref, issueKeys), 
+						createPullRequestUrl: generateCreatePullRequestUrl(pullRequest?.head?.repo?.html_url, pullRequest?.head?.ref, issueKeys || undefined),
```
Comment (KAbakumovAtl) [inline src/transforms/pull-request.ts:85]: I wonder if we should explicitly do issueKeys || undefined ? 

Looks like an excessive check. (Otherwise we'd better move it inside the function)

**Review by KAbakumovAtl — DISMISSED** [review]

File: src/transforms/util/pullRequestLinkGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Generates a create pull request link for GH in the format [URL]/compare/[name]?title=[...keys]-[name]
+ * This format allows the use of title query param to set the PR name
+ */
+export const generateCreatePullRequestUrl = (baseUrl: string, name: string, issueKeys: null | string[] = []) => {
```
Comment (ccrolf) [inline src/transforms/util/pullRequestLinkGenerator.ts:5]: Is it possible to use a feature-flag in this method, so we can do a gradual rollout (ie, reduce risk)?

Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts:5]: Can do, created flag use-new-github-pull-request-url-format, will include.

File: src/sync/transforms/branch.ts [hunk 8]
```diff
@@ -13,16 +14,13 @@ import _ from "lodash";
...
 	const branchKeys = issueKeyParser().parse(branch.name) || [];
 	const pullRequestKeys = issueKeyParser().parse(
 		branch.associatedPullRequests.nodes.length ? branch.associatedPullRequests.nodes[0].title : ""
 	) || [];
 	const commitKeys = issueKeyParser().parse(branch.target.message) || [];
-
-	const allKeys = branchKeys
-		.concat(pullRequestKeys)
-		.concat(commitKeys)
+	const allKeys = _.union(branchKeys, pullRequestKeys, commitKeys)
```
Comment (joshkay10) [inline src/sync/transforms/branch.ts:23]: Not sure what the feeling is around _ here, it was available so I used it.... can use straight JS if thats preferred?

This line is to remove duplicates.

File: src/sync/branches.ts [hunk 9]
```diff
@@ -13,6 +13,7 @@ export default async (logger: LoggerWithTarget, github: GitHubAPI, newGithub: Gi
 	logger.info("Syncing branches: started");
 
 	const useNewGHClient = await booleanFlag(BooleanFlags.USE_NEW_GITHUB_CLIENT_FOR_BRANCHES, false, jiraHost);
+	const useNewGHPrUrl = await booleanFlag(BooleanFlags.USE_NEW_GITHUB_PULL_REQUEST_URL_FORMAT, false, jiraHost);
```
Comment (joshkay10) [inline src/sync/branches.ts:16]: This was placed here instead of inside the transform/branches.ts files so I had access to Jira host, but also there was a bunch of extra conversions down the chain to async just to use the flag..... 
TLDR; was trying to keep the tidy up simpler.

File: src/sync/transforms/branch.ts (hunk 8 above)
Comment (KAbakumovAtl) [reply src/sync/transforms/branch.ts:23]: I don't have an opinion about it. Maybe @mboudreau can suggest?

**Review by KAbakumovAtl — APPROVED** [review]

File: src/transforms/util/pullRequestLinkGenerator.ts (hunk 5 above)
Comment (mboudreau) [reply src/transforms/util/pullRequestLinkGenerator.ts:9]: You still need to `encodeURIComponent` of name since a branch name can have `/` in it.

Comment (mboudreau) [reply src/transforms/util/pullRequestLinkGenerator.ts:9]: also, don't you need `quick_pull=1` in the querystring to make it a PR right away?  because the first page here is just a compare.  By adding this in the query, you go directly to the create PR page.

File: src/transforms/util/pullRequestLinkGenerator.ts (hunk 7 above)
Comment (mboudreau) [inline src/transforms/util/pullRequestLinkGenerator.ts:5]: issueKeys should never be `null` if  you're using a default value for it: `issueKeys: string[] = []`

File: src/sync/branches.ts [hunk 10]
```diff
@@ -36,6 +37,6 @@ export default async (logger: LoggerWithTarget, github: GitHubAPI, newGithub: Gi
 
 	return {
 		edges,
-		jiraPayload: transformBranches({ branches, repository })
+		jiraPayload: await transformBranches({ branches, repository, useNewGHPrUrl })
```
[the remark this answers is not in the document]
Comment (mboudreau) [reply src/sync/branches.ts:40]: hm, actually, you could separate each value into it's own attribute with typings to make it more succint.

**Review by mboudreau — CHANGES_REQUESTED** [review]
Noticed something missing in the generated URL

File: src/transforms/util/pullRequestLinkGenerator.ts (hunk 7 above)
Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts:5]: it was to appease typescript for the function issueKeyParser() in src/transforms/pull-request.ts which returns a null | string issuekey set,  but I can handle earlier before this functions called.

File: src/transforms/util/pullRequestLinkGenerator.ts (hunk 5 above)
Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts:9]: I have the return encoding the entire string 
	return  ``${baseUrl}/compare/${name}?title=${title}`;` that should cover it?

Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts]: appended quick_pull=1 too, can confirm it goes direct to PR.

Comment (KAbakumovAtl) [reply src/transforms/util/pullRequestLinkGenerator.ts]: But if we encode the entire string we also encode `/` and '?' so the URL will become invalid.

We need to encode each parameter individually. (The reason to encode the string to prevent injection attacks, so nobody will be able to put special characters in URL and make it pointing to another path)

Comment (KAbakumovAtl) [reply src/transforms/util/pullRequestLinkGenerator.ts]: So it should be
`${ encodeURIComponent(baseUrl)}/compare/${ encodeURIComponent(name)}?title=${ encodeURIComponent(title)}`

Comment (joshkay10) [reply src/transforms/util/pullRequestLinkGenerator.ts]: I accept, i have encoded each individually

	const title = encodeURIComponent(keys + name);
	const branchName = encodeURIComponent(name);
	const base = encodeURIComponent(baseUrl);
	const url = `${base}/compare/${branchName}?title=${title}&quick_pull=1`

**Review by KAbakumovAtl — DISMISSED** [review]

**Review by mboudreau — DISMISSED** [review]
Looking good

**Review by KAbakumovAtl — APPROVED** [review]
